mathics/builtin/arithmetic.py

```diff
@@ -133,7 +134,11 @@ def apply(self, z, evaluation):
                 elif mpmath.isnan(result):
                     result = Symbol("Indeterminate")
                 else:
-                    result = from_mpmath(result)
+                    try:
```

Comment:
Recently we had another "try except" the other day over something similar. 

The suggestion was to write a context handler that will handle this kind of stuff so that we not only do not duplicate boilerplate code like this, but we can customize it for things like tracing calls to SymPy (and here mpmath), get timings on that and so on. 

While that or this commit in of itself is not the end of the world, what has happened over time has that this kind of thing has gone on to the point where things have gotten pretty messy and while not gaining on general tools analogous to the benchmark stuff that could help everyone.

Reply:
@rocky, this is just what I meant when I say that each line we add to the builtins is affected by the refactor of the core. I just added something that I hoped was straightforward, but at the second like I hit this issue. 

Indeed, we could try here an approach using a context manager attached to the evaluation object, but that would imply touching the core. The best that I could do here is to add an inline comment talking about that idea.

Reply:
Thanks for asking in advance.

Adding a context manager to wrap calls to sympy or mpmath It doesn't change how evaluation works and is I don't see how any of the contemplated changes to evaluation would have anything to do with this.  

In fact I suspect all of these context managers could be in the mathics.builtin.base module. Changes in the evaluator have largely been in mathics.core.

Boxing evaluation and code on the other hand is totally different. Boxing is a kind of evaluation and has to be intertwined with evaluation. I am sorry I haven't fully explained my vision for how things might work. 

(The short answer would be something like do it like a number of other bytecode evaluators and if you can mix in JIT afterwards as Jan suggested, great. But this is very handwavy and I although I don't have a clearer sense exactly of how things would go. But I think I have a much clearer sense than others here.)

Reply:
> Thanks for asking in advance.
> 
> Adding a context manager to wrap calls to sympy or mpmath It doesn't change how evaluation works and is I don't see how any of the contemplated changes to evaluation would have anything to do with this.
> 

If I understood right, the context manager should be attached to the evaluation object. Again, we can do that, but I still need to figure out how to implement it, and it does not seem to be as easy as adding a try-except block. The good thing, in this case, is that I do not need to patch many modules, but just ``mathics.core.arithmetic``, where the superclasses of numeric builtins are defined.

> In fact I suspect all of these context managers could be in the mathics.builtin.base module. Changes in the evaluator have largely been in mathics.core.
> 

Maybe it is possible, but your proposal was to attach it to the evaluation object. 

> Boxing evaluation and code on the other hand is totally different. Boxing is a kind of evaluation and has to be intertwined with evaluation. I am sorry I haven't fully explained my vision for how things might work.
> 

Yes, I have a similar feeling, but let's discuss it when we reach it.

Reply:
> > Thanks for asking in advance.
> > Adding a context manager to wrap calls to sympy or mpmath It doesn't change how evaluation works and is I don't see how any of the contemplated changes to evaluation would have anything to do with this.
> 
> If I understood right, the context manager should be attached to the evaluation object. Again, we can do that, but I still need to figure out how to implement it, and it does not seem to be as easy as adding a try-except block. The good thing, in this case, is that I do not need to patch many modules, but just `mathics.core.arithmetic`, where the superclasses of numeric builtins are defined.
> 
> > In fact I suspect all of these context managers could be in the mathics.builtin.base module. Changes in the evaluator have largely been in mathics.core.
> 
> Maybe it is possible, but your proposal was to attach it to the evaluation object.

It would probably be simpler for me to just do it sometime.